python-checks/src/main/java/org/sonar/python/checks/StringFormatMisuseCheck.java

```diff
@@ -0,0 +1,135 @@
+import org.sonar.plugins.python.api.tree.BinaryExpression;
+import org.sonar.plugins.python.api.tree.DictionaryLiteral;
+import org.sonar.plugins.python.api.tree.DictionaryLiteralElement;
+import org.sonar.plugins.python.api.tree.KeyValuePair;
+import org.sonar.plugins.python.api.tree.StringLiteral;
+import org.sonar.plugins.python.api.tree.Tree;
+import org.sonar.plugins.python.api.tree.Tuple;
+
+@Rule(key = "S2275")
+public class StringFormatMisuseCheck extends PythonSubscriptionCheck {
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.MODULO, ctx -> {
+      BinaryExpression expression = (BinaryExpression) ctx.syntaxNode();
+      if (!expression.leftOperand().is(Tree.Kind.STRING_LITERAL)) {
+        return;
+      }
+
+      StringLiteral formatString = ((StringLiteral) expression.leftOperand());
+      Optional<StringFormat> formatOptional = StringFormat.createFromPrintfStyle(ctx, formatString, formatString.trimmedQuotesValue());
+
+      if (!formatOptional.isPresent()) {
+        // The string format contains invalid syntax.
+        return;
+      }
+
+      StringFormat format = formatOptional.get();
+      if (expression.rightOperand().is(Tree.Kind.TUPLE)) {
+        checkTuples(ctx, format, ((Tuple) expression.rightOperand()));
+      } else if (expression.rightOperand().is(Tree.Kind.DICTIONARY_LITERAL)) {
+        checkDictionaries(ctx, format, ((DictionaryLiteral) expression.rightOperand()));
+      } else if (format.numExpectedArguments() == 1) {
+        format.replacementFields().get(0).validateArgument(expression.rightOperand());
+      }
+    });
+  }
+
+  private static void checkTuples(SubscriptionContext ctx, StringFormat format, Tuple tuple) {
+    if (format.hasNamedFields()) {
+      ctx.addIssue(tuple, "Replace this formatting argument with a mapping.");
+      return;
+    }
+
+    if (format.numExpectedArguments() != tuple.elements().size()) {
+      reportInvalidArgumentSize(ctx, tuple, format.numExpectedArguments(), tuple.elements().size());
+      return;
+    }
+
+    for (int i = 0; i < tuple.elements().size(); ++i) {
+      format.replacementFields().get(i).validateArgument(tuple.elements().get(i));
+    }
+  }
+
+  private static void checkDictionaries(SubscriptionContext ctx, StringFormat format, DictionaryLiteral dict) {
+    if (format.hasPositionalFields()) {
+      ctx.addIssue(dict, "Replace this formatting argument with a tuple.");
+      return;
+    }
+
+    if (dict.elements().stream().anyMatch(StringFormatMisuseCheck::isOutOfScopeDictionaryElement)) {
+      // Do not bother with dictionaries containing unpacking expressions or keys which are not string literals.
+      return;
+    }
+
+    Map<String, List<StringFormat.ReplacementField>> fieldMap = format.replacementFields().stream().collect(
+      Collectors.groupingBy(StringFormat.ReplacementField::mappingKey));
+    if (fieldMap.size() != dict.elements().size()) {
```

Comment:
I believe we only want to detect missing arguments in the case of a dictionary. The following case doesn't fail, so I don't think it's part of the scope of this rule:
```
print("Hello %(first_name)s" % {"first_name": "John", "last_name": "Doe"})
```